### PR TITLE
more catch up

### PR DESCRIPTION
**Pull request by jimmyp**
What I've done since we last paired:
- competes `convert_to_typescript2.ts`
- completes `immutability.spec.ts`
- completes `typescriptquiz.ts` madness section
- adds multiple images and single state to web app
- uses ` setState(prev => ...)` in web app

What I still need to look at when I get some time:
- couldn't get exhaustive pattern matching working in `convert_to_typescript2.ts`
- still have a bug in my image swapping code, just need to work through the index arithmetic carefully when I get some time

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,28 +1,71 @@
...
 
-const Image = (props: {rotation: number}) =>
-  <img
+type State = {
+  imageIndexOffset: number
+  imagesWithRotations: { url: string, rotation: number}[]
+}
+
+function Image(props: {state: State}){
+  const currentImage = props.state.imagesWithRotations[getCurrentImageIndex(props.state)];
```
Comment (hackle) [inline src/App.tsx]: recommend only putting what's required in the `props` in this case the current image

File: src/App.tsx [hunk 2]
```diff
@@ -1,28 +1,71 @@
...
+  return state.imageIndexOffset % state.imagesWithRotations.length;
+}
 
-const Controls = (props: { onCounterClockwiseClick: () => void, onClockwiseClick: () => void }) =>
+function rotateCurrentImage(rotationDelta: number, prev: State): State
+{
+  const currentImageIndex = getCurrentImageIndex(prev);
+
+  //TODO: this code still has a bug in it
+  const newImages = [
```
Comment (hackle) [inline src/App.tsx]: this works, recommend trying out  `images.map((img, index) => ...)`  to be more succinct

File: src/convert_to_typescript2.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    type: 'PLAY'
+}
+
+type Events = Eat | Sleep | Work
+
+function assertUnreachable(x: never): never {
+    throw new Error("Didn't expect to get here");
+}
+
+//TODO: Couldn't get exhaustive pattern matching working?
```
Comment (hackle) [inline src/convert_to_typescript2.ts:28]: the trick here is to use string for `event.type`, e.g. 
```
type Eat = { type: 'Eat', ... }
```

File: src/immutability.spec.ts [hunk 4]
```diff
@@ -72,14 +81,19 @@ describe('Immutability quiz!', () => {
             }
         ];
 
-        const withCorrectCapital: Person = chris;
-        withCorrectCapital.favourites.food = expectedFavourites;
+        const newFavouriteFoods: Person = {
+            ...chris,
+            favourites: {
+                ...chris.favourites,
+                food: [...chris.favourites.food.filter(f => f.name !== 'sushi'), baoBun]
```
Comment (hackle) [inline src/immutability.spec.ts:88]: this is great - to be more succinct again we can use `map`

File: src/convert_to_typescript2.ts (hunk 3 above)
Comment (jimmyp) [reply src/convert_to_typescript2.ts:28]: I've changed to string as you suggested, this code does work if you add `Play` to the union. But I get a false negative when the union is just `Eat | Sleep | Work`. The compiler error I get in this case is:

```
TypeScript error in /Users/jim/code/callery/src/convert_to_typescript2.ts(34,36):
Argument of type 'any' is not assignable to parameter of type 'never'.  TS2345

    32 |         case 'SLEEP': return `You will dream about ${event.dream} for ${event.duration} hours`;
    33 |         case 'WORK': return `A job in ${event} is tough but you can always ${event.slack}!`;
  > 34 |         default: assertUnreachable(event.type)
       |                                    ^
    35 |     }
    36 | }
    37 | console.log(handleEvent({ type: 'EAT', food: 'Pizza' }));

```

Maybe we can look at this as part of our review?

**Pull request merged**